Allow OHOS servoshell to have a simple multiple tab implementation.

Narfinger opened the pull request:
Currently we just pause the compositor and replace the window in it
while having separate bookkeeping to remember which window belongs to
which tab.
Currently there are no tests for OHOS, so we cannot test the changes.

jschwe commented on ports/servoshell/egl/ohos.rs:
```diff
@@ -126,6 +130,11 @@ static PROMPT_TOAST: OnceLock<
 
+/// Currently we do not support different contexts for different windows but we might want to change tabs.
+/// For this we store the window context for every tab and change the compositor by hand.
+static WEBVIEW_TO_RAW_HANDLE: LazyLock<Mutex<Vec<(XComponentWrapper, WindowWrapper)>>> =
```
Why do we need the LazyLock here? both Mutex and Vec should be able to be constructed in `const` contexts.

jschwe commented on ports/servoshell/egl/ohos.rs:
```diff
@@ -613,6 +662,12 @@ pub fn init_servo(init_opts: InitOpts) -> napi_ohos::Result<()> {
 
+#[napi]
+fn focus_webview(id: u32) {
+    error!("Focusing from napi {}", id);
```
```suggestion
    debug!("Focusing from napi {}", id);
```

jschwe commented on ports/servoshell/egl/ohos/resources.rs:
```diff
@@ -12,6 +12,8 @@ pub(crate) struct ResourceReaderInstance {
 
 impl ResourceReaderInstance {
     pub(crate) fn new(resource_dir: PathBuf) -> Self {
+        use log::error;
+        error!("Resource dir {:?}", resource_dir);
```
Is this leftover from debugging?

jschwe reviewed (commented):
There are still a couple of `error` log messages in the code which you probably used during debugging and some newly added commented code. Could you do another pass and check what can be removed or downgraded to `debug` / `trace` log messages?

Narfinger committed:
Allow OHOS servoshell to have a simple multiple tab implementation.

Currently we just pause the compositor and replace the window in it
while having separate bookkeeping to remember which window belongs to
which tab.

Signed-off-by: Narfinger <Narfinger@users.noreply.github.com>

Narfinger commented:
Forgot the cleanup. Thanks for noticing.

jschwe commented on ports/servoshell/egl/ohos.rs:
```diff
@@ -220,50 +253,60 @@ extern "C" fn on_surface_created_cb(xcomponent: *mut OH_NativeXComponent, window
-    // Each thread will send its id via the channel
-    let _main_surface_thread = thread::spawn(move || {
-        let (tx, rx): (Sender<ServoAction>, Receiver<ServoAction>) = mpsc::channel();
+    if !SERVO_CHANNEL.get().is_some() {
```
This looks racy to me. You could use one of the `Once` or `Lazy` types

Narfinger replied:
I don't quite understand, the SERVO_CHANNEL is a OnceLock already.

jschwe commented on ports/servoshell/egl/ohos.rs:
```diff
@@ -613,6 +656,12 @@ pub fn init_servo(init_opts: InitOpts) -> napi_ohos::Result<()> {
 
+#[napi]
+fn focus_webview(id: u32) {
+    error!("Focusing from napi {}", id);
```
```suggestion
    debug!("Focusing webview {id} from napi");
```

jschwe commented on support/openharmony/entry/src/main/ets/pages/Index.ets:
```diff
@@ -78,57 +86,87 @@ struct Index {
+            } else {
+              this.tablist.push(this.tablist[this.tablist.length-1]+1);
+            }
+            // yes this is correct as we always have one tab extra
```
Why do we have one extra tab?

Narfinger replied:
The extra tab is not in the tablist but a separate ArkUI element. This makes the initialization easier.

Narfinger committed:
Fix a crash in the touch handler.
If some move events take the long route over the script thread but get
cleaned up by a fling the late one will crash the handler.
Now we just ignore move events that are already cleaned up.

Signed-off-by: Narfinger <Narfinger@users.noreply.github.com>

Narfinger committed:
Incorporate suggestions.

Co-authored-by: Jonathan Schwender <55576758+jschwe@users.noreply.github.com>
Signed-off-by: Narfinger <Narfinger@users.noreply.github.com>

Narfinger commented on ports/servoshell/egl/app_state.rs:
```diff
@@ -163,6 +163,11 @@ impl WebViewDelegate for RunningAppState {
+                .url()
+                .map(|u| u.to_string())
+                .unwrap_or(String::from("about:blank"));
+            self.callbacks.host_callbacks.on_url_changed(url);
```
I changed this part but I am not sure if it is better this way.

mrobinson commented on components/compositing/webview_renderer.rs:
```diff
@@ -395,6 +395,7 @@ impl WebViewRenderer {
         self.dispatch_input_event(event);
     }
 
+    /// Send touch event to the pipeline. Returns true if the event was send
```
```suggestion
    /// Send a [`TouchEvent`] to the Constellation for this [`WebViewRenderer`]. 
    /// Returns true if the event was sent and false otherwise.
```

mrobinson commented on ports/servoshell/egl/app_state.rs:
```diff
@@ -337,6 +337,19 @@ impl RunningAppState {
         self.inner_mut().webviews.insert(webview.id(), webview);
     }
 
+    pub(crate) fn activate_webview(&self, id: u32) {
```
```suggestion
    pub(crate) fn focus_webview(&self, id: WebViewId) {
```

Narfinger replied:
I changed the name of the function but kept the id for now.

mrobinson replied:
I think it makes sense to pass `WebViewId` and to store it everywhere in order to make the type of the id clear.

Narfinger replied:
Ok I changed it now and I think this makes more sense. One snag that I experienced was that the active webview changes quite slowly in RunningAppState. For this reason we now need access to the newest_webview.

mrobinson reviewed (approved):
The non-OHOS parts seem fine, though we aren't sure if we are going to continue to expose `WebView`. In that case, this code will need to change. Please wait for @jschwe to review the OHOS bits.

jschwe commented on components/compositing/touch.rs:
```diff
@@ -374,81 +378,88 @@ impl TouchHandler {
+        // sequence take the long route over the script thread and
+        // can arrive after we cleaned them up from a fling event.
+        // We can ignore these as these are already cleaned up
+        if let Some(touch_sequence) = self.try_get_current_touch_sequence_mut() {
```
outdated, sorry

Narfinger commented on support/openharmony/entry/src/main/ets/pages/Index.ets:
```diff
@@ -78,57 +86,89 @@ struct Index {
+            })
+        }.tabBar('1')
+        ForEach(this.tablist, (item: number) => {
+          TabContent() {
```
TabContent() are the contents of a tab so two tabs would look like `Tabs() TabContent() { foo}.tabBar('FooTab') TabContent() { bar }.tabBar('BarTab')`.
The ForEach allows us to iterate over an array and fill the UI with multiple of the next TabContent. However, we want to always have one tab and because this tab needs the initialization code, we split it out of the loop. We also do not have it in the tablist because of the same reason.

Narfinger commented on support/openharmony/entry/src/main/ets/pages/Index.ets:
```diff
@@ -78,57 +86,89 @@ struct Index {
+          TabContent() {
+            XComponent(this.xComponentAttrs)
+              .focusable(true)
+          }.tabBar(String(item))
```
To me it looks like there is no problem, however, I could not test every site. We use the same `get_xcomponent_size` and `get_xcomponent_offset` methods.

jschwe commented on components/shared/embedder/input_events.rs:
```diff
@@ -124,6 +124,7 @@ pub enum TouchEventType {
 pub struct TouchId(pub i32);
 
 /// An ID for a sequence of touch events between a `Down` and the `Up` or `Cancel` event.
+/// The ID is the same for all events between `Down`` and `Up`` or `Cancel``
```
```suggestion
/// The ID is the same for all events between `Down` and `Up` or `Cancel`
```

jschwe commented on ports/servoshell/egl/ohos.rs:
```diff
@@ -185,6 +203,56 @@ impl ServoAction {
+                    .url()
+                    .map(|u| u.to_string())
+                    .unwrap_or(String::from("about:blank"));
+                SET_URL_BAR_CB
```
I think I would modify the callback to have both the webview id (From servo) and the URL as parameters. 
On the ArkTS side, you could keep a map of servo_webview_id -> url.
